src/core/engine/ffmpeg/ffmpegreplaygain.cpp

```diff
@@ -0,0 +1,383 @@
+    }
+};
+using FilterContextPtr = std::unique_ptr<AVFilterContext, FilterContextDeleter>;
+
+struct FilterGraphDeleter
+{
+    void operator()(AVFilterGraph* graph) const
+    {
+        if(graph) {
+            avfilter_graph_free(&graph);
+        }
+    }
+};
+using FilterGraphPtr = std::unique_ptr<AVFilterGraph, FilterGraphDeleter>;
+
+struct FilterInOutDeleter
+{
+    void operator()(AVFilterInOut* inout) const
+    {
+        if(inout) {
+            avfilter_inout_free(&inout);
+        }
+    }
+};
+using FilterInOutPtr = std::unique_ptr<AVFilterInOut, FilterInOutDeleter>;
+
+struct ReplayGainResult
+{
+    double gain;
+    double peak;
+};
+
+struct ReplayGainFilter
+{
+    AVFilterContext* filterContext;
+    AVFilterInOut* filterOutput;
+    FilterGraphPtr filterGraph;
+};
+
+constexpr auto FrameFlags = AV_BUFFERSRC_FLAG_KEEP_REF | AV_BUFFERSRC_FLAG_NO_CHECK_FORMAT | AV_BUFFERSRC_FLAG_PUSH;
+
+bool finishFilter(AVFilterContext* filter)
+{
+    return av_buffersrc_add_frame_flags(filter, nullptr, FrameFlags) >= 0;
+}
+
+ReplayGainResult extractRGValues(AVFilterGraph* graph)
+{
+    ReplayGainResult result{};
+
+    av_opt_get_double(graph->filters[1], "integrated", AV_OPT_SEARCH_CHILDREN, &result.gain);
+    av_opt_get_double(graph->filters[1], "true_peak", AV_OPT_SEARCH_CHILDREN, &result.peak);
+    result.gain = -18.0 - result.gain; // TODO - newer standard uses 23.0 as reference. make configurable?
+    result.peak = std::pow(10, result.peak / 20.0);
+
+    return result;
+}
+} // namespace
+
+namespace Fooyin {
+class FFmpegReplayGainPrivate
+{
+public:
+    explicit FFmpegReplayGainPrivate(MusicLibrary* library);
+
+    [[nodiscard]] bool setupTrack(const Track& track, ReplayGainFilter& filter);
+    [[nodiscard]] ReplayGainResult handleTrack(bool inAlbum) const;
+    void handleAlbum(const TrackList& album);
+
+    AudioFormat m_format;
+    FormatContextPtr m_formatCtx;
+    CodecContextPtr m_decoderCtx;
+    ReplayGainFilter m_albumFilter;
+    ReplayGainFilter m_trackFilter;
+    int m_streamIndex{-1};
+
+    MusicLibrary* m_library;
+};
+
+FFmpegReplayGain::FFmpegReplayGain(MusicLibrary* library, QObject* parent)
+    : Worker{parent}
+    , p{std::make_unique<FFmpegReplayGainPrivate>(library)}
+{ }
+
+FFmpegReplayGain::~FFmpegReplayGain() = default;
+
+FFmpegReplayGainPrivate::FFmpegReplayGainPrivate(MusicLibrary* library)
+    : m_albumFilter{}
+    , m_trackFilter{}
+    , m_library{library}
+{ }
+
+void FFmpegReplayGain::calculate(const TrackList& tracks, bool asAlbum)
+{
+    setState(Running);
+
+    if(asAlbum) {
+        p->handleAlbum(tracks);
+    }
+    else {
+        for(auto track : tracks) {
+            if(p->setupTrack(track, p->m_trackFilter)) {
+                const auto result = p->handleTrack(false);
+                track.setRGTrackGain(static_cast<float>(result.gain));
+                track.setRGTrackPeak(static_cast<float>(result.peak));
+                p->m_library->writeTrackMetadata({track});
```

Comment:
Same as the for the comment in handleAlbum, this should be moved outside of the loop and a single call made.